pkg/summarizer/flakiness.go

```diff
@@ -0,0 +1,286 @@
+		totalRuns:          totalCount,
+		totalRunsWithInfra: totalCountWithInfra,
+		passedRuns:         test.passed,
+		failedRuns:         test.failed,
+		failedInfraRuns:    test.failedInfraCount,
+		flakyRuns:          test.flakyCount,
+		infraInfo:          infraInfo,
+	}
+	return testInfo, true
+
+}
+
+func createHealthiness(startDate int, endDate int, results []Result, testByEnv map[string]TestInfo, infraIssues map[string]int) Healthiness {
+	healthiness := Healthiness{
+		startDate:   startDate,
+		endDate:     endDate,
+		tests:       []TestInfo{},
+		infraIssues: make(map[string]int),
+	}
+
+	// The Go way of sorting the keys of a map in descending order
+	var keys []int
+	for k := range HEALTHY_RANGE {
+		keys = append(keys, k)
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(keys)))
+	for _, threshold := range keys {
+		newBucket := FlakyBucket{
+			threshold: float64(threshold),
+		}
+		healthiness.flakyBuckets = append(healthiness.flakyBuckets, newBucket)
+	}
+
+	averageFlakiness := 0.0
+	for _, testInfo := range testByEnv {
+		healthiness.tests = append(healthiness.tests, testInfo)
+		averageFlakiness += testInfo.flakiness
+		for i, bucket := range healthiness.flakyBuckets {
+			if testInfo.flakiness > bucket.threshold {
+				healthiness.flakyBuckets[i].tests += 1.0
+				break
+			}
+		}
+	}
+	healthiness.totalTests = len(healthiness.tests)
+	healthiness.totalConfigs = len(results)
+	if healthiness.totalTests > 0 {
+		healthiness.averageFlakiness = averageFlakiness / float64(healthiness.totalTests)
+	}
+	for k, v := range infraIssues {
+		healthiness.infraIssues[k] = v
+	}
+	return healthiness
+}
+
+func calculateInfraInfo(issues map[string]int, failedCount int) string {
+	result := make([]string, 0)
+	if len(issues) > 0 && failedCount > 0 {
+		// Sorts the map items by value (int) and then key (string) if values are equal
+		// The sort is in descending order: [5,4,3]
+		items := make([]IntString, 0)
+		for key, value := range issues {
+			items = append(items, IntString{
+				s: key,
+				i: value,
+			})
+		}
+		sort.Slice(items, func(i, j int) bool {
+			// These two comparisons enforce descending order for the integers
+			if items[i].i > items[j].i {
+				return true
+			}
+			if items[i].i < items[j].i {
+				return false
+			}
+			// String comparison is still ascending: [a,b,c]
+			return sortorder.NaturalLess(items[i].s, items[j].s)
+		})
+		for _, item := range items {
+			result = append(result, item.s+fmt.Sprintf(" %.2f%% ", 100*float64(item.i)/float64(failedCount)))
+		}
+	}
+	return strings.TrimSpace(strings.Join(result, ""))
+}
+
+func categorizeFailure(resultCounts *Result, message string) {
+	if message == "" || !INFRA_REGEX.MatchString(message) {
+		resultCounts.failed += 1
+		return
+	}
+	resultCounts.failedInfraCount += 1
+	resultCounts.infraFailures[message] += 1
+}
+
+func getValueOfFlakyResult(resultCounts *Result) {
+	// Default behavior of adding a 50% flakiness
+	flakiness := DEFAULT_FLAKINESS
```

Comment:
(how does this going to work? I assume each flaky cell has different flakiness data?)

Reply:
Maybe it can be an element of the TestGroup Config? Or Per-Tab?

Reply:
Is this attempting to weigh how 'flaky' a flaky result is, e.g. how many attempts failed out of the total? TestGrid should have that info at least, though I don't know how difficult that is to grab from the stored state (I think you'd need to hack it at the moment from the short_text/icon of the status).

Reply:
IMO If it's too much work, I'd suggest we add proper support later (add a TODO), and leave it as 50% passing for now. GKE did not enable retry on flakes so I'm not too worry about this.

Reply:
I ended up doing just that (the TODO), and I'll get to it later.